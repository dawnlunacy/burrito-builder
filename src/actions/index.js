export const setOrders = orders => ({
  type: 'SET_ORDERS',
  orders
});

export const handleError = errorMessage => ({
  type: 'SET_ERROR_MESSAGE',
  errorMessage
});