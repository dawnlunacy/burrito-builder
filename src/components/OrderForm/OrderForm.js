import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { setOrders, handleError } from '../../actions';
import { getOrders } from '../../apiCalls';



export class OrderForm extends Component {
  constructor() {
    super();
    this.state = {
      id: Date.now(),
      name: '',
      ingredients: []
    };
  }

  componentDidMount() {
    const { setOrders } = this.props
    console.log("SET", setOrders)
    getOrders()
      .then(data => setOrders(data.orders))
      .catch(err => console.error('Error fetching:', err));
      
    
  }

  handleNameChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleIngredientChange = e => {
    e.preventDefault();
    this.setState({ingredients: [...this.state.ingredients, e.target.name]});
  }

  handleSubmit = e => {
    e.preventDefault();
    this.checkInputs();
  }

  checkInputs = () => {
    const { handleError, setOrders, orders } = this.props;
    if (this.state.ingredients.length === 0) {
      handleError(`${this.state.name} Please select some ingredients`)
    }
    else {
      setOrders([...orders, this.state]);
      this.clearInputs();
    }
  }

  clearInputs = () => {
    this.setState({name: '', ingredients: []});
  }


  render() {
    const {errorMessage } = this.props;
    const possibleIngredients = ['beans', 'steak', 'carnitas', 'sofritas', 'lettuce', 'queso fresco', 'pico de gallo', 'hot sauce', 'guacamole', 'jalapenos', 'cilantro', 'sour cream'];
    const ingredientButtons = possibleIngredients.map(ingredient => {
      return (
        <button key={ingredient} name={ingredient} onClick={e => this.handleIngredientChange(e)}>
          {ingredient}
        </button>
      )
    });

    return (
      <form>
        <input
          type='text'
          placeholder='Name'
          name='name'
          value={this.state.name}
          onChange={e => this.handleNameChange(e)}
        />

        { ingredientButtons }

        <p>Order: { this.state.ingredients.join(', ') || 'Nothing selected' }</p>
        <p> { errorMessage } </p>
        
        <button onClick={e => this.handleSubmit(e)}>
          Submit Order
        </button>
      </form>
    )
  }
}

// export default OrderForm;

export const mapStateToProps = state => ({
  orders: state.orders,
  errorMessage: state.errorMessage
});

export const mapDispatchToProps = dispatch => (
  bindActionCreators({
    setOrders,
    handleError
  }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(OrderForm);