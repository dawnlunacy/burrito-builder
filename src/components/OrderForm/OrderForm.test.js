import React from 'react';
import { OrderForm, mapDispatchToProps, mapStateToProps } from './OrderForm';
import { shallow } from 'enzyme';
import { setOrders } from '../../actions/index';

describe('OrderForm', () => {
  let wrapper;

  const mockSetOrders = jest.fn();
  const mockHandleError = jest.fn();
  const mockErrorMessage = "Please select some ingredients";

  const mockOrders = [{
    id: 0,
    name: "Lacy",
    ingredients: ["AwesomeSauce", "Optimism"]
  }, 
  {
    id: 1,
    name: "Robbie",
    ingredients: ["Pants", "Pugs", "Sweaters"]
  }];
  

  beforeEach(() => {
    wrapper = shallow(<OrderForm 
      setOrders={mockSetOrders}
      orders={mockOrders}
      handleError={mockHandleError}
      errorMessage={mockErrorMessage}

    />);

  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state of when handleNameChange is called with an event', () => {
    const event = {
      target: {
        name: "name",
        value: "Trappis"
      }
    }

    wrapper.instance().handleNameChange(event);

    expect(wrapper.state().name).toEqual(event.target.value);
    });

    it('should set state of when handleIngredientChange is called with an event', () => {
      const event = {
        preventDefault: jest.fn(),
        target: {
          name: "queso fresco",
        }
      }
  
      wrapper.instance().handleIngredientChange(event);
  
      expect(wrapper.state().ingredients[0]).toEqual(event.target.name);
      })

  describe('mapStateToProps', () => {
    it('it should return an object with orders and error message', () => {

      const mockState = {
        orders: mockOrders,
        errorMessage: ''
      };

      const expected = {
        orders: mockOrders,
        errorMessage: ''
      }

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected)
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls setOrders when updateOrders is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = mockOrders;

      const actionToDispatch = setOrders(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setOrders(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it.skip('calls handleError when checkInputs is called from handleIngredientChange without a value', () => {
      wrapper.instance().handleError = jest.fn();
      wrapper.instance().checkInputs = jest.fn();
      wrapper.instance().handleIngredientChange = jest.fn();
      const mockEvent = {
        preventDefault: jest.fn(),
        target: {
          name: "queso fresco",
        }
      }
      // const mockState 
      wrapper.find('button').at(5).simulate('click', mockEvent);

      expect(wrapper.instance().handleIngredientChange).toHaveBeenCalled();
      expect(wrapper.instance().checkInputs).toHaveBeenCalled();
      expect(wrapper.instance().handleError).toHaveBeenCalledWith(mockErrorMessage);
    });

    it('should run handleSubmit on click', () => {
      wrapper.instance().handleSubmit = jest.fn();

      const mockEvent = { preventDefault: jest.fn() }

      wrapper.find('button').at(12).simulate('click', mockEvent);

      expect(wrapper.instance().handleSubmit).toHaveBeenCalledWith(mockEvent);
    });
  });

});