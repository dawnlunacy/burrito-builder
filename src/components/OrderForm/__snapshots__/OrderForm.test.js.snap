// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm should match the snapshot 1`] = `
<form>
  <input
    name="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value=""
  />
  <button
    key="beans"
    name="beans"
    onClick={[Function]}
  >
    beans
  </button>
  <button
    key="steak"
    name="steak"
    onClick={[Function]}
  >
    steak
  </button>
  <button
    key="carnitas"
    name="carnitas"
    onClick={[Function]}
  >
    carnitas
  </button>
  <button
    key="sofritas"
    name="sofritas"
    onClick={[Function]}
  >
    sofritas
  </button>
  <button
    key="lettuce"
    name="lettuce"
    onClick={[Function]}
  >
    lettuce
  </button>
  <button
    key="queso fresco"
    name="queso fresco"
    onClick={[Function]}
  >
    queso fresco
  </button>
  <button
    key="pico de gallo"
    name="pico de gallo"
    onClick={[Function]}
  >
    pico de gallo
  </button>
  <button
    key="hot sauce"
    name="hot sauce"
    onClick={[Function]}
  >
    hot sauce
  </button>
  <button
    key="guacamole"
    name="guacamole"
    onClick={[Function]}
  >
    guacamole
  </button>
  <button
    key="jalapenos"
    name="jalapenos"
    onClick={[Function]}
  >
    jalapenos
  </button>
  <button
    key="cilantro"
    name="cilantro"
    onClick={[Function]}
  >
    cilantro
  </button>
  <button
    key="sour cream"
    name="sour cream"
    onClick={[Function]}
  >
    sour cream
  </button>
  <p>
    Order: 
    Nothing selected
  </p>
  <p>
     
    Please select some ingredients
     
  </p>
  <button
    onClick={[Function]}
  >
    Submit Order
  </button>
</form>
`;
